File: type_erasure/cpp/te_map.cpp
#include <any>
#include <iostream>
#include <string>
#include <unordered_map>

struct MyClass
{
    auto print() -> void
    {
        std::cout << "MyClass prints!\n";
    }
};

struct TypeErasedMap
{
    template< typename T >
    void store( std::string const& name, T& obj )
    {
        storage[name] = &obj;
    }

    template< typename T >
    auto get( std::string const& name ) -> T&
    {
        auto iter = storage.find( name );

        if( iter != storage.end() ) {
            if( auto* ptr = std::any_cast< T* >( iter->second ) ) {
                return *ptr;
            }
        }

        throw std::runtime_error( "obj not found" );
    }

private:
    std::unordered_map< std::string, std::any > storage;
};

auto main() -> int
{
    TypeErasedMap map;
    MyClass mc;
    map.store( "MyClass", mc );
    auto& ref = map.get< MyClass >( "MyClass" );
    ref.print();
    return 0;
}

File: type_erasure/rs/te_map.rs
use std::any::Any;
use std::collections::HashMap;

struct MyClass;
impl MyClass {
    fn print(&self) {
        println!("MyClass prints!");
    }
}

struct TypeErasedMap<'a> {
    storage: HashMap<String, &'a dyn Any>,
}

impl<'a> TypeErasedMap<'a> {
    fn new() -> Self {
        Self {
            storage: HashMap::new(),
        }
    }

    fn store<T: Any>(&mut self, name: &str, obj: &'a T) {
        self.storage.insert(name.to_string(), obj);
    }

    fn get<T: Any>(&self, name: &str) -> Option<&T> {
        self.storage
            .get(name)
            .and_then(|any| any.downcast_ref::<T>())
    }
}

fn main() {
    let mut map = TypeErasedMap::new();
    let mc = MyClass;

    map.store("MyClass", &mc);

    if let Some(ref_mc) = map.get::<MyClass>("MyClass") {
        ref_mc.print();
    }
}
